tests/test_monitor_system.py

```diff
@@ -0,0 +1,205 @@
+            f"{METRIC_DISK_TOTAL_GB}/main": Union(int, float),
+            f"{METRIC_DISK_TOTAL_GB}/home": Union(int, float),
+        }
+    )
+
+
+def test_monitor_system_is_false(tmp_dir, mocker):
+    mock_psutil(mocker)
+    with Live(
+        tmp_dir,
+        save_dvc_exp=False,
+        monitor_system=False,
+    ) as live:
+        assert live._cpu_monitor is None
+
+
+def test_monitor_system_is_true(tmp_dir, mocker):
+    mock_psutil(mocker)
+    with Live(
+        tmp_dir,
+        save_dvc_exp=False,
+        monitor_system=True,
+    ) as live:
+        cpu_monitor = live._cpu_monitor
+        assert isinstance(live._cpu_monitor, CPUMonitor)
+        # start is called but end not called
+        assert cpu_monitor._shutdown_event._flag is False
+
+    # end was called
+    assert cpu_monitor._shutdown_event._flag is True
+
+
+def test_ignore_non_existent_directories(tmp_dir, mocker):
+    mock_psutil(mocker)
+    with Live(
+        tmp_dir,
+        save_dvc_exp=False,
+        monitor_system=False,
+    ) as live:
+        non_existent_disk = "/non-existent"
+        monitor = CPUMonitor(
+            disks_to_monitor={"main": "/", "non-existent": non_existent_disk}
+        )
+        monitor(live)
+        metrics = monitor._get_metrics()
+        monitor.end()
+
+    assert not Path(non_existent_disk).exists()
+
+    assert f"{METRIC_DISK_USAGE_PERCENT}/non-existent" not in metrics
+    assert f"{METRIC_DISK_USAGE_GB}/non-existent" not in metrics
+    assert f"{METRIC_DISK_TOTAL_GB}/non-existent" not in metrics
+
+
+def test_monitor_system_metrics(tmp_dir, mocker):
+    mock_psutil(mocker)
+    interval = 0.05
+    num_samples = 4
+    with Live(
+        tmp_dir,
+        save_dvc_exp=False,
+        monitor_system=False,
+    ) as live:
+        live.cpu_monitor = CPUMonitor(interval=interval, num_samples=num_samples)
+        time.sleep(interval * num_samples + interval)  # log metrics once
+        live.next_step()
+        time.sleep(interval * num_samples + interval)  # log metrics twice
+        live.make_summary()
+
+        _, latest = parse_metrics(live)
+
+    schema = {}
+    for name in [
+        "step",
+        METRIC_CPU_COUNT,
+        METRIC_CPU_USAGE_PERCENT,
+        METRIC_CPU_PARALLELIZATION_PERCENT,
+        METRIC_RAM_USAGE_PERCENT,
+        METRIC_RAM_USAGE_GB,
+        METRIC_RAM_TOTAL_GB,
+        f"{METRIC_DISK_USAGE_PERCENT}/main",
+        f"{METRIC_DISK_USAGE_GB}/main",
+        f"{METRIC_DISK_TOTAL_GB}/main",
+    ]:
+        dpath.new(schema, name, Union(int, float))
+
+    assert latest == S(schema)
+
+
+def test_monitor_system_timeseries(tmp_dir, mocker):
+    mock_psutil(mocker)
+    interval = 0.05
+    num_samples = 4
+    with Live(
+        tmp_dir,
+        save_dvc_exp=False,
+        monitor_system=False,
+    ) as live:
+        live.cpu_monitor = CPUMonitor(interval=interval, num_samples=num_samples)
+        time.sleep(interval * num_samples + interval)  # log metrics once
```

Comment:
this is still unreliable, I would do a while loop with a small sleep interval that reads from disk or checks some Live vat to detect an update

to test that it respects intervals, etc - mock sleep, run thread itself from the test and check the result

Reply:
just some ideas ^^ that might be a better way, but we can't go with unreliable / flakey tests

Reply:
I tried to respect what you explained. If there is still some misunderstanding between us, could you please refer to a link or a concrete example to illustrate your point? This way, I might be able to understand better what you meant :)